src/tike/ptycho/solvers/rpie.py

```diff
@@ -322,4 +384,68 @@ def _get_nearplane_gradients(
         psi_update_denominator,
         probe_update_numerator,
         probe_update_denominator,
+        position_update_numerator,
+        position_update_denominator,
     )
+
+
+def _mad(x, **kwargs):
+    """Return the mean absolute deviation around the median."""
+    return cp.mean(cp.abs(x - cp.median(x, **kwargs)), **kwargs)
+
+
+def _update_position(
+    scan,
+    position_options,
+    position_update_numerator,
+    position_update_denominator,
+    alpha=0.05,
+    max_shift=1,
+):
+    step = position_update_numerator / (
+        (1 - alpha) * position_update_denominator +
+        alpha * position_update_denominator.max()
+    )
+
+    step_x = step[..., 0]
+    step_y = step[..., 1]
+
+    if position_options.use_adaptive_moment:
+        logger.info(
+            "position correction with ADAptive Momemtum acceleration enabled.")
+        step_x, position_options.vx, position_options.mx = adam(
+            step_x,
+            position_options.vx,
+            position_options.mx,
+            vdecay=position_options.vdecay,
+            mdecay=position_options.mdecay)
+        step_y, position_options.vy, position_options.my = adam(
+            step_y,
+            position_options.vy,
+            position_options.my,
+            vdecay=position_options.vdecay,
+            mdecay=position_options.mdecay)
+
+    # Step limit for stability
+    _max_shift = cp.minimum(
+        max_shift,
+        _mad(
+            cp.concatenate((step_x, step_y), axis=-1),
+            axis=-1,
+            keepdims=True,
+        ),
+    )
+    step_x = cp.maximum(-_max_shift, cp.minimum(step_x, _max_shift))
+    step_y = cp.maximum(-_max_shift, cp.minimum(step_y, _max_shift))
+
+    # SYNCHRONIZE ME?
```

Comment:
Does this comment belong in the final merge?

Reply:
I tried to replace this zero net movement constraint with regularization as we discussed last week, but I could not tune the regularization so that it would keep the reconstructed image from drifting. The regularization was either too strong or not strong enough! Thus, I'm back to this zero net movement constraint. What I don't know is if this constraint needs to be synchronized across devices. I guess I need to do some testing today.